SecretManager/src/V1/Client/BaseClient/SecretManagerServiceBaseClient.php

```diff
@@ -0,0 +1,674 @@
+    private const SERVICE_ADDRESS = 'secretmanager.googleapis.com';
+
+    /** The default port of the service. */
+    private const DEFAULT_SERVICE_PORT = 443;
+
+    /** The name of the code generator, to be included in the agent header. */
+    private const CODEGEN_NAME = 'gapic';
+
+    /** The default scopes required by the service. */
+    public static $serviceScopes = [
+        'https://www.googleapis.com/auth/cloud-platform',
+    ];
+
+    private static function getClientDefaults()
+    {
+        return [
+            'serviceName' => self::SERVICE_NAME,
+            'apiEndpoint' => self::SERVICE_ADDRESS . ':' . self::DEFAULT_SERVICE_PORT,
+            'clientConfig' => __DIR__ . '/../../resources/secret_manager_service_client_config.json',
+            'descriptorsConfigPath' => __DIR__ . '/../../resources/secret_manager_service_descriptor_config.php',
+            'gcpApiConfigPath' => __DIR__ . '/../../resources/secret_manager_service_grpc_config.json',
+            'credentialsConfig' => [
+                'defaultScopes' => self::$serviceScopes,
+            ],
+            'transportConfig' => [
+                'rest' => [
+                    'restClientConfigPath' => __DIR__ . '/../../resources/secret_manager_service_rest_client_config.php',
+                ],
+            ],
+        ];
+    }
+
+    /**
+     * Formats a string containing the fully-qualified path to represent a project
+     * resource.
+     *
+     * @param string $project
+     *
+     * @return string The formatted project resource.
+     */
+    public static function projectName($project)
+    {
+        return self::getPathTemplate('project')->render([
+            'project' => $project,
+        ]);
+    }
+
+    /**
+     * Formats a string containing the fully-qualified path to represent a secret
+     * resource.
+     *
+     * @param string $project
+     * @param string $secret
+     *
+     * @return string The formatted secret resource.
+     */
+    public static function secretName($project, $secret)
+    {
+        return self::getPathTemplate('secret')->render([
+            'project' => $project,
+            'secret' => $secret,
+        ]);
+    }
+
+    /**
+     * Formats a string containing the fully-qualified path to represent a
+     * secret_version resource.
+     *
+     * @param string $project
+     * @param string $secret
+     * @param string $secretVersion
+     *
+     * @return string The formatted secret_version resource.
+     */
+    public static function secretVersionName($project, $secret, $secretVersion)
+    {
+        return self::getPathTemplate('secretVersion')->render([
+            'project' => $project,
+            'secret' => $secret,
+            'secret_version' => $secretVersion,
+        ]);
+    }
+
+    private static function registerPathTemplates()
+    {
+        self::loadPathTemplates(__DIR__ . '/../../resources/secret_manager_service_descriptor_config.php', self::SERVICE_NAME);
+    }
```

Comment:
Did we consider using the `descriptorsConfigPath` in `getClientDefaults` instead, and putting `registerPathTemplates` directly in `ResourceHelperTrait`? It would look something like:

```php
trait ResourceHelperTrait
{
    /**
     * placeholder for this function like we have in GapicClientTrait
     */
    private static function getClientDefaults()
    {
        return [];
    }

    private static function registerPathTemplates()
    {
        $templateConfigPath = self::getClientDefaults()['descriptorsConfigPath'];
        self::loadPathTemplates($templateConfigPath, self::SERVICE_NAME);
    }
}
```

Reply:
We hadn't that's a good idea. If `self::SERVICE_NAME` is private, can the `ResourceHelperTrait` still use it?

Reply:
Yes because trait methods are treated as if they were copy/pasted directly in the class itself

Reply:
Ok! Another question, can two traits used on the same client define the same method? If both `ResourceHelperTrait` and `GapicClientTrait` define a `getClientDefaults()` will they collide during trait resolution? Or are both just ignored b.c the client `use`'ing the traits defines an overload?

Reply:
https://github.com/googleapis/gax-php/pull/447

Reply:
I just pushed an update to this PR with a local fix to remove the `registerPathTemplates` from the GAPIC (to go along with the PR above)

Reply:
Resolving as this will be fixed with my pending batch of changes.